File: tcg-java/src/de/kimminich/kata/tcg/Move.java
package de.kimminich.kata.tcg;

import java.util.Optional;

public class Move {

    private final Optional<Card> card;
    private final Action action;

    public Move(Optional<Card> card, Action action) {
        this.card = card;
        this.action = action;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        Move move = (Move) o;

        if (action != move.action) return false;
        if (card != null ? !card.equals(move.card) : move.card != null) return false;

        return true;
    }

    @Override
    public int hashCode() {
        int result = card != null ? card.hashCode() : 0;
        result = 31 * result + (action != null ? action.hashCode() : 0);
        return result;
    }

    public Optional<Card> getCard() {
        return card;
    }
}

File: tcg-kotlin/src/main/kotlin/de/kimminich/kata/tcg/Move.kt
package de.kimminich.kata.tcg

class Move(val card: Card, val action: Action) {

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as Move

        if (card != other.card) return false
        if (action != other.action) return false

        return true
    }

    override fun hashCode(): Int {
        var result = card.hashCode()
        result = 31 * result + action.hashCode()
        return result
    }

    override fun toString(): String {
        return "Move(card=$card, action=$action)"
    }


}
